### PR TITLE
feat(builder): broccoli bump to 1.0

### PR DESCRIPTION
**Pull request by hansl**
This is the right PR for the broccoli upgrade.

Original was https://github.com/angular/angular-cli/pull/299.

### CODE REVIEW

**Comment by jkuri** [comment]
Cool! I will close mine PR but leave local branch if you find anything useful.

File: lib/broccoli/angular2-app.js [hunk 1]
```diff
@@ -41,33 +41,31 @@ Angular2App.prototype.toTree = function() {
...
+        // Recursively call this function with the full sub-path.
+        fs.readdirSync(filePath).forEach(function(n) {
+          addPathRecursive(path.join(name, n));
+        });
+      }
+    });
 
+    // Because the tsconfig does not include the source directory, add this as the first path
+    // element.
+    tsconfig.files = tsconfig.files.map(name => path.join(sourceDir, name));
```
Comment (filipesilva) [inline lib/broccoli/angular2-app.js]: In line 52, this is already being done for test files. Won't those test files end up having `sourceDir` twice at the start?

**Comment by filipesilva** [comment]
I went over the changes and it lgtm. Also checked the CI logs and couldn't find any weird stuff.

**Comment by filipesilva** [comment]
Setting `"compileOnSave": false,` on `tsconfig.json` makes every save break on my IDE (Atom). 

I thought it might be a sign that there's something wrong with manual compilation. Compiling with `tsc -p` works, but compiling directly from the `src` dir does not.

```
filipe.silva@DESKTOP-7ND6T3R MINGW64 /E/dev/ideablade/testing/testproj2 (master)
$ tsc -p src/

filipe.silva@DESKTOP-7ND6T3R MINGW64 /E/dev/ideablade/testing/testproj2 (master)
$ cd src/

filipe.silva@DESKTOP-7ND6T3R MINGW64 /E/dev/ideablade/testing/testproj2/src (master)
$ tsc
app.ts(1,25): error TS2307: Cannot find module 'angular2/platform/browser'.
app/testproj2.ts(1,25): error TS2307: Cannot find module 'angular2/core'.
app/testproj2.ts(2,64): error TS2307: Cannot find module 'angular2/router'.
```

**Comment by filipesilva** [comment]
Running `ng build` will generate `.js` and `.js.map` files inside `src/app`. @jkuri and I were able to reproduce this.

File: lib/broccoli/broccoli-typescript.js [hunk 2]
```diff
@@ -1,170 +1,333 @@
...
+ * Broccoli plugin that implements incremental Typescript compiler.
+ *
+ * It instantiates a typescript compiler instance that keeps all the state about the project and
+ * can re-emit only the files that actually changed.
+ *
+ * Limitations: only files that map directly to the changed source file via naming conventions are
+ * re-emitted. This primarily affects code that uses `const enum`s, because changing the enum value
+ * requires global emit, which can affect many files.
+ */
+class DiffingTSCompiler extends Plugin {
```
Comment (jkuri) [inline lib/broccoli/broccoli-typescript.js]: :+1: nice

File: lib/broccoli/broccoli-typescript.js [hunk 3]
```diff
@@ -1,170 +1,333 @@
...
+        this.tsServiceHost = new CustomLanguageServiceHost(
+          this.tsOpts, this.rootFilePaths, this.fileRegistry, this.inputPaths[0]);
+        this.tsService = ts.createLanguageService(this.tsServiceHost, ts.createDocumentRegistry());
+    }
 
-  filesToRemove.forEach(fileName => {
-    this.removeFile(fileName, this.inputPaths[0], this.outputPath);
-  });
-};
+    build() {
```
Comment (jkuri) [inline lib/broccoli/broccoli-typescript.js]: I would use `this.listEntries()` in `build()` method, this is new in broccoli 1.0 and it simplifies things a lot. 
With that method you get times when file was last changed and compare them to previous times, if there is no time in object yet it means it is the initial build. Thats how I did it. Not saying it is the best solution but it is definitely less complex and even more faster. 

Please see https://github.com/jkuri/angular-cli/blob/broccoli-1.0beta/lib/broccoli/broccoli-typescript.js#L32-#L65

**Comment by hansl** [comment]
@filipesilva I fixed the bug where `js` and `js.map` were generated in the source directory. Everything should be fine now.

File: lib/broccoli/broccoli-typescript.js (hunk 3 above)
Comment (hansl) [inline lib/broccoli/broccoli-typescript.js]: `listEntries()` and `listFiles()` isn't what we're looking for here though. These will include _all_ source files.

I'll have to add some extra logic for this to work. I admit it'll be ideal, but surprisingly right now this is really fast. Let's just wait for a refactor of broccoli-typescript which will be the next PR.

**Comment by filipesilva** [comment]
Tested and can confirm it's fixed.

File: lib/broccoli/broccoli-typescript.js (hunk 3 above)
Comment (jkuri) [inline lib/broccoli/broccoli-typescript.js]: Okay, I agree. Let me check this now and then I can give my lgtm.

File: lib/broccoli/angular2-app.js [hunk 4]
```diff
@@ -41,33 +41,31 @@ Angular2App.prototype.toTree = function() {
...
-    // TODO(i): kill rootFilePaths in broccoli-typescript and use tsconfig.json#files instead
-    tsConfigCompilerOptions.rootFilePaths = tsConfigCompilerOptions.rootFilePaths.
-        concat(tsconfig.files.map(function(p) {
-          // TODO(i): this is a hack - for some reason we need to prefix all paths with srcDir because
-          //          tsc's "rootDir" doesn't take effect when resolving "files" paths
-          return path.join(sourceDir, p);
-        }));
+    // Add all spec files to files. We need this because spec files are their own entry
+    // point.
+    fs.readdirSync(sourceDir).forEach(function addPathRecursive(name) {
```
Comment (jkuri) [inline lib/broccoli/angular2-app.js]: :100:

File: lib/broccoli/angular2-app.js [hunk 5]
```diff
@@ -113,10 +111,10 @@ Angular2App.prototype.toTree = function() {
       jsTree,
       this.index(),
       vendorNpmTree,
-      thirdPartyJs
+      thirdPartyJs,
```
Comment (jkuri) [inline lib/broccoli/angular2-app.js]: can you remove this please, we are not using it anymore?

Comment (hansl) [inline lib/broccoli/angular2-app.js]: Done.

**Comment by hansl** [comment]
Also added a test to make sure `ng build` doesn't add new files to git.

**Comment by jkuri** [comment]
I still need to bug a bit, unfortunately there is still issue when some `.ts` file changes;
`file changed app/proj1.ts
ENOENT: no such file or directory, scandir '/Users/jan/Desktop/angular-cli/proj1/tmp/funnel-input_base_path-tOrNtqOJ.tmp/0/src/'
`
After this will be solved it is ready for merge, I will help you check it. 

Btw, you increased performance by 1/3 :+1:

**Comment by jkuri** [comment]
I cannot comment on the lines that are not being changed so I will write fixes here;
- in `broccoli-typescript.js:115` add `o.name = path.join(_this.outputPath, o.name);`
- in `service-worker-manifest.js:40` replace with `var manifestPath = path.join(this.cachePath, MANIFEST_FILE)`

Please let me know when you will be done so I can check it.

**Comment by hansl** [comment]
Fixed what I think were all bugs. Please take another look.

@alxhub: could you take a look at the SWManifest? I stopped reading the manifest file and instead keep the hash in memory.

**Comment by jkuri** [comment]
LGTM

**Pull request merged**